### ISSUE TITLE
Exception while using with circe

### ISSUE THREAD

**Issue by dantodor** [gnieh/diffson#49]
I have the following json:
```json
          {
            "order": {
            "items": [
                  {
                    "type": "Item1",
                    "payload": [
                      {
                        "id": 123,
                        "description": "banana",
                        "quantity": 1
                      },
                      {
                        "id": 456,
                        "description": "apple",
                        "quantity": 2
                      }
                    ]
                  },
                  {
                    "type": "Item2",
                    "payload": [
                      {
                        "id": 456,
                        "name": "apple"
                      }
                    ]
                  }
                ]
            }
          }
```

I'm just evaluating if I can use Diffson with Circe for deep comparing json files. In the code I have 
`val patch = JsonDiff.diff(json,json2,false)`

where the difference between the two  is that I invert **banana** and **apple** items in the **Item1** payload.
When running the above I'm getting the following exception:

```
Exception in thread "main" java.lang.ClassCastException: scala.collection.immutable.Vector cannot be cast to scala.collection.immutable.List
	at gnieh.diffson.CirceInstance$provider$$anonfun$unapplyArray$1.apply(CirceInstance.scala:180)
	at scala.Option.map(Option.scala:146)
	at gnieh.diffson.CirceInstance$provider$.unapplyArray(CirceInstance.scala:180)
	at gnieh.diffson.CirceInstance$provider$.unapplyArray(CirceInstance.scala:142)
	at gnieh.diffson.JsonSupport$JsonProvider$JsArray$.unapply(JsonProvider.scala:74)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.diff(JsonDiff.scala:58)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.fields$1(JsonDiff.scala:91)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.fieldsDiff(JsonDiff.scala:101)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.diff(JsonDiff.scala:57)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.fields$1(JsonDiff.scala:91)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.fieldsDiff(JsonDiff.scala:101)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.diff(JsonDiff.scala:57)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.diff(JsonDiff.scala:47)
	at gnieh.diffson.JsonDiffSupport$JsonDiff.diff(JsonDiff.scala:41)
```

NB, Circe is able to parse the json without issues.

Using scala 2.11.8, sbt 0.13.8, circe 0.7.0, include Diffson as advised in the readme.
What am I doing wrong ?

**Comment by satabin** [comment]
Hi,

This exception looks weird indeed. Can you please provide `json2` as well so that I can try on my computer? Thanks.

**Comment by agourlay** [comment]
This exception is due to you having circe 0.7.0 in your classpath.
Since Circe 0.7.0 the JSON array are represented by Vectors instead of List.
  
As the current release of diffson is depending on Circe 0.6.x, you are hitting a binary incompatibility.

**Comment by satabin** [comment]
Ok so I will release the 2.1.2 version. @dantodor can you give a try to diffson version `2.1.2-SNAPSHOT`to see if it solves your problem?

**Comment by dantodor** [comment]
Here is the second : `{
            "order": {
            "items": [
                  {
                    "type": "Item1",
                    "payload": [
                      {
                        "id": 456,
                        "description": "apple",
                        "quantity": 2
                      },
                      {
                        "id": 123,
                        "description": "banana",
                        "quantity": 1
                      }
                    ]
                  },
                  {
                    "type": "Item2",
                    "payload": [
                      {
                        "id": 456,
                        "name": "apple"
                      }
                    ]
                  }
                ]
            }
          }`

I'll give it a try right away ... Maven or checkout from here ?

**Comment by satabin** [comment]
The last snapshot version on maven should contain the merge of #44.

**Comment by dantodor** [comment]
with Circe 0.6.2 works, I'll try the snapshot later... However, here is the result with third parameter as true:

`[
  {
    "op" : "add",
    "path" : "/order/items/0/payload/0",
    "value" : {
      "id" : 456,
      "description" : "apple",
      "quantity" : 2
    }
  },
  {
    "op" : "remove",
    "path" : "/order/items/0/payload/2",
    "old" : {
      "id" : 456,
      "description" : "apple",
      "quantity" : 2
    }
  }
]`

Probably not in specs, but for me the two should be (arguably) equivalent.  Or have an extra parameter to tell that I want non-positional array equivalence testing. Like map. If I get some slack ofo of work, I hope I can look into this.

**Comment by satabin** [comment]
Yes, the `old` field is not in the RFC but should be ignored by RFC implementations. It is a conservative extension specific to diffson.

Non-positional array equivalence will be hard to implement, because Json arrays are ordered and adding this to diffson would make generated patches incompatible with other implementations.

**Comment by dantodor** [comment]
Well, back to square one then ... make a big map and deep compare :)... Thanks for your help!

**Comment by satabin** [comment]
Or you can use diffson to compare your maps :)

**Comment by dantodor** [comment]
I think it's somewhere in the middle :) ... Map compare works well for equality:
`val m1 = Map(1->2,3->Map(1->2,3->4))
  val m2 = Map(1->2,3->Map(3->4,1->2))

  println(m1==m2)
`
As for finding differences... another story.
I think I'll use both ... To give you a broader picture, I'm trying to verify the json output from some functions and check it against well-known json. So, it might be the case of a two-pronged approach, if map equality fails, do the patch via Diffson....
BTW, how would one use Diffson on maps ?

**Comment by satabin** [comment]
Hi,

I just released diffson 2.1.2 and uploaded artifacts to sonatype maven repositories.

Diffson can compute diffs for all types that it knows how to convert into Json based on some json library dependent conversion protocol. If you teach circe how to convert your map into a valid json value, then you can simply make a `JsonDiff.diff(map1, map2, false)`.
